### DISCUSSION TITLE
Different Get-Member output after piping object array to Select-Object

### DISCUSSION THREAD

**Discussion by ghost** [PowerShell/PowerShell#18122] [Q&A]
I've seeing a confusing behavior in regards to `Select-Object`. Hoping someone can explain it to me.

1. Declare an array of PSCustomObjects
`$Data = [PSCustomObject]@{Name = 'Bob'}, [PSCustomObject]@{Name = 'Susan'}`

2. Run `$Data | Get-Member` and it shows one result
```
   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
Name        NoteProperty string Name=Bob
```

3. Run `$Data | Select-Object -First 1` 
4. Run `$Data | Get-Member` again, and it now shows **two** results. Why would piping the array to `Select-Object` cause the output of `$Data | Get-Member` to change? This happens on both 5.1 and 7.2.
```
   TypeName: Selected.System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
Name        NoteProperty string Name=Bob

   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
Name        NoteProperty string Name=Susan
```

**Comment by rhubarb-geek-nz** [comment]
Interesting, when I repeated your steps I got the same answer.

However...

```
#!/usr/bin/env pwsh

$Data = [PSCustomObject]@{Name = 'Bob'}, [PSCustomObject]@{Name = 'Susan'}

Write-Host ( $Data )

Write-Host ( $Data | Select-Object -First 1 )

Write-Host ( $Data )
```

this outputs

```
@{Name=Bob} @{Name=Susan}
@{Name=Bob}
@{Name=Bob} @{Name=Susan}
```

So the Select-Object didn't really change the contents of the array.

**Comment by rhubarb-geek-nz** [comment]
Now my theory is that PowerShell is throwing around real C# objects and they have internal state. So perhaps an array has the list of actual elements, and also has a selected item. When the list is created the selected item is null, but when you use Select-Object, it then selects the object.

If you look at the output, the first item is

TypeName: Selected.System.Management.Automation.PSCustomObject

the 2nd is 

System.Management.Automation.PSCustomObject

I am going to suggest two things, 

1. Get-Member just asks the object for whatever it has and tries its best to interpret it

2. The pipe operator is not like a pipe in UNIX shell, it is actually passing a real object to the next Cmdlet. So the Select-Object can affect the internal state, and indeed that is the purpose of Select-Object. In other words, yes, the input is not read-only.

Well that's my theory.
